From: SJ Park <sj@kernel.org>
Date: Thu, 30 Jul 2026 07:27:45 -0700
Subject: [PATCH] mm/damon/core: fixup last_psi_total initialization

Initializing to zero at allocation makes first diff too big.  Set it to
the real total PSI at the beginning of kdamond.

Signed-off-by: SJ Park <sj@kernel.org>
---
 mm/damon/core.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index ee49eda8384d1..3a66c35ef1209 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -705,8 +705,6 @@ struct damos_quota_goal *damos_new_quota_goal(
 	goal->metric = metric;
 	goal->target_value = target_value;
 	INIT_LIST_HEAD(&goal->list);
-	if (goal->metric == DAMOS_QUOTA_SOME_MEM_PSI_US)
-		goal->last_psi_total = 0;
 	return goal;
 }
 
@@ -4070,6 +4068,18 @@ static int kdamond_wait_activation(struct damon_ctx *ctx)
 	return -EBUSY;
 }
 
+static void damos_init_quota_goal_last_psi(struct damos *s)
+{
+	struct damos_quota_goal *goal;
+
+	damos_for_each_quota_goal(goal, &s->quota)
+	{
+		if (goal->metric != DAMOS_QUOTA_SOME_MEM_PSI_US)
+			continue;
+		goal->last_psi_total = damos_get_some_mem_psi_total();
+	}
+}
+
 static void kdamond_init_ctx(struct damon_ctx *ctx)
 {
 	unsigned long sample_interval = ctx->attrs.sample_interval ?
@@ -4087,6 +4097,7 @@ static void kdamond_init_ctx(struct damon_ctx *ctx)
 	damon_for_each_scheme(scheme, ctx) {
 		damos_set_next_apply_sis(scheme, ctx);
 		damos_set_filters_default_reject(scheme);
+		damos_init_quota_goal_last_psi(scheme);
 	}
 }
 
-- 
2.47.3

